src/plugins/keras_support/categorical_crossentropy_operation.cpp

```diff
@@ -0,0 +1,248 @@
+                return np.sum(target * -np.log(output), axis=-1, keepdims=False)
+        )")
+    };
+    const double small = 1e-7;
+
+    ///////////////////////////////////////////////////////////////////////////
+    cat_cross_operation::cat_cross_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    primitive_argument_type cat_cross_operation::cat_cross0d() const
+    {
+        return primitive_argument_type{static_cast<double>(1.)};
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    primitive_argument_type cat_cross_operation::cat_cross1d(arg_type&& target,arg_type&& output,bool from_logits) const
+    {
+        if(from_logits)
+        {
+            output.vector() = blaze::softmax(output.vector());
+        }
+        else
+        {
+            output.vector() /= blaze::sum(output.vector());
+        }
+
+        // Construct low and high clip regions
+        auto sz0 = output.dimension(0);
+        primitive_argument_type lo_{small}, hi_{1-small};
+
+        auto lo = extract_value_vector<double>(lo_, sz0, name_, codename_ );
+        auto hi = extract_value_vector<double>(hi_, sz0, name_, codename_ );
+
+        output.vector() = blaze::max(blaze::min(output.vector(), hi.vector()), lo.vector());
+
+        auto res = target.vector() * (-blaze::log(output.vector()));
+        return primitive_argument_type{ blaze::sum(res) };
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+
+    primitive_argument_type cat_cross_operation::cat_cross2d(
+        arg_type&& target, arg_type&& output,bool from_logits) const
+    {
+        if(from_logits)
+        {
+            output.matrix() = blaze::softmax(output.matrix());
+        }
+        else
+        {
+            output.matrix() /= blaze::sum(output.matrix());
+        }
+
+        // Construct low and high clip regions
+        auto sz0 = output.dimension(0);
+        auto sz1 = output.dimension(1);
+        primitive_argument_type lo_{small}, hi_{1-small};
+
+        auto lo = extract_value_matrix<double>(lo_, sz0, sz1, name_, codename_ );
+        auto hi = extract_value_matrix<double>(hi_, sz0, sz1, name_, codename_ );
+
+        output.matrix() = blaze::max(blaze::min(output.matrix(), hi.matrix()), lo.matrix());
+
+        auto res = target.matrix() % (-blaze::log(output.matrix()));
+        return primitive_argument_type{ blaze::sum(res) };
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
+
+    primitive_argument_type cat_cross_operation::cat_cross3d(
+        arg_type&& target, arg_type&& output, bool from_logits) const
+    {
+        if(from_logits)
+        {
+            output.tensor() = blaze::softmax(output.tensor());
```

Comment:
I think this does not work. `blaze::softmax` calculates `softmax` on the whole tensor, while Keras softmax calculates softmax where `axis=-1` in its default case. In the given NumPy implementation, Keras softmax is used. We have a softmax implementation in Phylanx which imitates its corresponding functionality in Keras (calculates on `axis=-1` where no axis is given). That is why I suggested categorical cross-entropy to be added to the softmax primitive.

Reply:
As an example
```py
import keras.backend as K
x = K.constant([[[1, 2, 3, 4],[5, 6, 7, 8],[9, 10, 11, 12]],
       [[-1, -2, -3, -4], [-5, -6, -7, -8], [-9, -10, -11, -12]]])
a = K.softmax(x)
K.get_value(a)
```
```log
array([[[0.0320586 , 0.08714432, 0.2368828 , 0.6439142 ],
        [0.0320586 , 0.08714432, 0.2368828 , 0.6439142 ],
        [0.0320586 , 0.08714432, 0.2368828 , 0.6439142 ]],

       [[0.6439142 , 0.2368828 , 0.08714432, 0.0320586 ],
        [0.6439142 , 0.2368828 , 0.08714432, 0.0320586 ],
        [0.6439142 , 0.2368828 , 0.08714432, 0.0320586 ]]], dtype=float32)
```
and
```c++
blaze::DynamicTensor<int> t{
        {{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}},
        {{-1, -2, -3, -4}, {-5, -6, -7, -8}, {-9, -10, -11, -12}}};

std::cout << blaze::softmax(t) << std::endl;
```
```log
(( 1.05575e-05  2.86983e-05    7.801e-05  0.000212053 ) (  0.00057642   0.00156687    0.0042592    0.0115777 ) (   0.0314715    0.0855484     0.232545     0.632122 ) )
((  1.4288e-06  5.25627e-07  1.93367e-07  7.11359e-08 ) ( 2.61694e-08   9.6272e-09  3.54165e-09   1.3029e-09 ) (  4.7931e-10  1.76328e-10  6.48676e-11  2.38634e-11 ) )
```